vault/core.go

```diff
@@ -1185,21 +1198,20 @@ func (c *Core) unsealPart(ctx context.Context, seal Seal, key []byte, useRecover
 
 			// We have recovery keys; we're going to use them as the new
 			// barrier key.
-			if err := c.barrier.Rekey(ctx, recoveryKey); err != nil {
-				return nil, errwrap.Wrapf("error rekeying barrier during migration: {{err}}", err)
+			err = c.seal.GetAccess().(*shamirseal.ShamirSeal).SetKey(recoveryKey)
+			if err != nil {
+				return nil, errwrap.Wrapf("failed to set master key in seal: {{err}}", err)
 			}
-
-			if err := c.barrier.Delete(ctx, StoredBarrierKeysPath); err != nil {
-				// Don't actually exit here as successful deletion isn't critical
-				c.logger.Error("error deleting stored barrier keys after migration; continuing anyways", "error", err)
+			if err := c.seal.SetStoredKeys(ctx, [][]byte{masterKey}); err != nil {
+				return nil, errwrap.Wrapf("error setting new barrier key information during migrate: {{err}}", err)
 			}
 
 			masterKey = recoveryKey
 
 		case c.seal.RecoveryKeySupported():
 			// The new seal will have recovery keys; we set it to the existing
 			// master key, so barrier key shares -> recovery key shares
-			if err := c.seal.SetRecoveryKey(ctx, masterKey); err != nil {
+			if err := c.seal.SetRecoveryKey(ctx, newRecoveryKey); err != nil {
```

Comment:
If we are moving from an new style shamir seal to a autoseal should we instead be using the old shared keys as the recovery key? (newRecoveryKey == master key here)

Reply:
We should, yes.  When we're migrating in unsealPart, `seal` is the migration seal. 
 So we took the else branch above on line 1146, setting `masterKey = recoveredKey`, i.e. the barrier unseal key that was passed in by the user.  Then we assigned `newRecoveryKey := masterKey`.  So from what I can tell we're doing what you suggest, as verified by TestSealMigration.

Reply:
Ah okay, you're right for some reason i thought `newRecoveryKey` was being set after we update `masterKey` to the stored key.